harness/determined/profiler.py

```diff
@@ -0,0 +1,517 @@
+        self.current_batch_idx = 0
+        self.agent_id = agent_id
+        self.trial_id = trial_id
+        self.master_url = master_url
+        self.start_on_batch = start_on_batch
+        self.end_after_batch = end_after_batch
+
+        # Set up timer to stop collecting after 5 minutes
+        self.has_started = False
+        self.has_finished = False
+        self.max_collection_seconds = 300
+        self.shutdown_timer = threading.Timer(self.max_collection_seconds, self._end_collection)
+        self.shutdown_timer.daemon = True
+        self.shutdown_lock = threading.Lock()
+
+        # Set up the thread responsible for making API calls
+        self.send_queue = queue.Queue()
+        self.sender_thread = ProfilerSenderThread(self.send_queue, self.master_url)
+        self.sender_thread.start()
+
+        # Launch the system metric collecting thread, but not the actual collection of metrics
+        self.sys_metric_collector_thread = SysMetricCollectorThread(trial_id, agent_id, self.send_queue)
+        self.sys_metric_collector_thread.start()
+
+        # TODO: Add data structure to batch timings and then send to SenderThread
+        #       Does this need to be its own thread to flush correctly?
+        # self.timings_batcher = TimingsBatcher()
+
+    @property
+    def is_enabled(self):
+        return self.has_started and not self.has_finished
+
+    def update_batch_idx(self, new_batch_idx: int):
+        self.current_batch_idx = new_batch_idx
+        self.sys_metric_collector_thread.update_batch_idx(self.current_batch_idx)
+
+        # Check if we should start collecting metrics
+        if not self.is_enabled and not self.has_finished and self.current_batch_idx >= self.start_on_batch:
+            self._begin_collection()
+
+        # Check if we should stop collecting metrics due to batch idx being exceeded
+        if self.is_enabled and self.end_after_batch is not None and self.current_batch_idx > self.end_after_batch:
+            self._end_collection()
+
+    def _begin_collection(self):
+        self.sys_metric_collector_thread.activate()
+        # TODO: Start up TimingBatcher as well
+        self.shutdown_timer.start()
+        self.has_started = True
+
+    def _end_collection(self):
+        """
+        Stop collecting data and shut down child threads. This function can be invoked due to the
+        max batch idx being exceeded or due to timeout, so the function needs to be threadsafe.
+        """
+        # Make _end_collection idempotent
+        with self.shutdown_lock:
+            if self.has_finished:
+                return
+
+            # Shut down in reverse creation order
+            self.sys_metric_collector_thread.kill()
+            self.sys_metric_collector_thread.join()
+
+            self.sender_thread.kill()
+            self.sender_thread.join()
+
+            self.has_finished = True
```

Comment:
Do you actually need the shutdown lock and the has_finished check?  I think the kill() calls and the join() are basically already idempotent (other than extra dummy messages getting put into queues... pretty harmless)

Reply:
I think using a single lock is cleaner, easier to reason about and it comes at no meaningful cost. `kill` doesn't need to be idempotent, that's just a benefit of how it's currently implemented, which could easily change in the future. I'd rather write code that handles the concurrency explicitly in the function that needs to be threadsafe rather than relying on implementation details of a function that has no reason to remain idempotent.

Reply:
Honestly I think any differences between cleanliness and ease of reasoning between either the lock solution or the queue solution are pretty tiny.

I guess that's not really what I'm concerned about.  I'll respond in more detail on the other comment.